corecfg/picfg.go

```diff
@@ -0,0 +1,146 @@
+ *
+ */
+
+package corecfg
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+)
+
+// valid pi config keys
+var piConfigKeys = map[string]bool{
+	"disable_overscan":         true,
+	"framebuffer_width":        true,
+	"framebuffer_height":       true,
+	"framebuffer_depth":        true,
+	"framebuffer_ignore_alpha": true,
+	"overscan_left":            true,
+	"overscan_right":           true,
+	"overscan_top":             true,
+	"overscan_bottom":          true,
+	"overscan_scale":           true,
+	"display_rotate":           true,
+	"hdmi_group":               true,
+	"hdmi_mode":                true,
+	"hdmi_drive":               true,
+	"avoid_warnings":           true,
+	"gpu_mem_256":              true,
+	"gpu_mem_512":              true,
+	"gpu_mem":                  true,
+	"sdtv_aspect":              true,
+	"config_hdmi_boost":        true,
+	"hdmi_force_hotplug":       true,
+}
+
+func updatePiConfig(path string, config map[string]string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// build regexp
+	cfgKeys := make([]string, len(config))
+	i := 0
+	for k := range config {
+		cfgKeys[i] = k
+		i++
+	}
+	reStr := fmt.Sprintf(`^[ \t]*?(?P<is_comment>#?)[ \t#]*?(?P<key>%s)=(?P<old_value>.*)$`, strings.Join(cfgKeys, "|"))
+	rx := regexp.MustCompile(reStr)
+
+	// now go over the content
+	found := map[string]bool{}
+	needsWrite := false
+	var toWrite []string
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		matches := rx.FindStringSubmatch(line)
+		if len(matches) > 0 {
+			wasComment := (matches[1] == "#")
+			key := matches[2]
+			oldValue := matches[3]
+			found[key] = true
+			if config[key] != "" {
+				if wasComment || oldValue != config[key] {
+					line = fmt.Sprintf("%s=%s", key, config[key])
+					needsWrite = true
+				}
+			} else {
+				line = fmt.Sprintf("#%s=%s", key, oldValue)
+				if !wasComment {
+					needsWrite = true
+				}
+
+			}
+		}
+		toWrite = append(toWrite, line)
+	}
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+
+	// write anything that is missing
+	for key := range config {
+		if !found[key] && config[key] != "" {
+			needsWrite = true
+			toWrite = append(toWrite, fmt.Sprintf("%s=%s", key, config[key]))
+		}
+	}
+
+	if needsWrite {
+		s := strings.Join(toWrite, "\n")
+		return osutil.AtomicWriteFile(path, []byte(s), 0644, 0)
+	}
+
+	return nil
+}
+
+func piConfig() string {
+	return filepath.Join(dirs.GlobalRootDir, "/boot/uboot/config.txt")
+}
+
+func handlePiConfiguration() error {
+	if osutil.FileExists(piConfig()) {
+		// snapctl can actually give us the whole dict in
+		// JSON, in a single call; use that instead of this.
+		config := map[string]string{}
+		for key := range piConfigKeys {
+			output, err := snapctlGet(fmt.Sprintf("pi-config.%s", strings.Replace(key, "_", "-", -1)))
+			if err != nil {
```

Comment:
It does not current: 
```
$ snap run --shell hello 
$ env|grep ^SNAP_NAME
SNAP_NAME=hello
$ snapctl get option-does-not-exists ; echo $?

0
$ snapctl get unexisting unexisting2 ; echo $?
{}
0
```
Maybe it should, this would make some things easier.
```